engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -184,12 +185,21 @@ public void afterExecutionPathCompleted(
     final Either<Failure, Boolean> canBeCompletedOrFailure =
         canBeCompletedWithCompletionCondition(element, flowScopeContext);
     if (canBeCompletedOrFailure.isLeft()) {
+      // this incident is un-resolvable
       incidentBehavior.createIncident(canBeCompletedOrFailure.getLeft(), flowScopeContext);
       return;
     } else if (canBeCompletedOrFailure.isRight() && canBeCompletedOrFailure.get()) {
-      // terminate all remaining child instances because the completion condition evaluates to true.
-      stateTransitionBehavior.terminateChildInstances(flowScopeContext);
-      if (loopCharacteristics.isSequential()) {
+      final var size =
+          stateBehavior.getChildInstances(flowScopeContext).stream()
+              .filter(child -> ProcessInstanceLifecycle.canTerminate(child.getIntent()))
+              .count();
+      if (size > 0) {
+        // terminate all remaining child instances because the completion condition evaluates to
+        // true.
+        stateTransitionBehavior.terminateChildInstances(flowScopeContext);
+      }
+
+      if (loopCharacteristics.isSequential() || size == 0) {
```

Comment:
At first, I really wanted to do this,  because our processor was executed asynchronously. I was worried that the number of the child instance would change after the appendFollowUpCommand was executed. Sorry, maybe I thought too much。

Reply:
@Lzgabel Please don't apologize for this. There's no need. 🙂 

That's actually a reasonable worry. Let me try to explain how it works:

Since version `1.0` we have a strong separation between commands and events. State can only be updated by writing events. When processed, commands can produce new commands and/or new events. State changes for an event take effect immediately when the event is written to the log. Commands however are postponed to when it's their turn to be processed. You can read more about it in [ZEP004](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md).

To come back to this situation, this method is called from a child processor when that child is being completed, or in other words: we're currently processing `ProcessInstance:COMPLETE_ELEMENT` command for what ever the child instance element is. During the processing of this complete-element command we can only write new commands or events. `appendFollowUpCommand` means we're writing a command, which will be processed in the future. No state is changed by this directly. Only once that command is processed we would likely write some events (like `ELEMENT_TERMINATED`) which will actually have effects on the state by reducing the number of child instances.

Does that make sense? Let me know if I can clarify anything. I'm happy to help out 👍

Reply:
Hi @korthout , thank you very much for patiently answering my confusion. 🤝